client/session.go

```diff
@@ -1429,17 +1526,12 @@ func (s *session) streamCollectedBlocksMetadata(
 		}
 
 		received, ok := metadata[m.id.Hash()]
-		if !ok {
+		if !ok || received.submitted {
 			received = &receivedBlocks{
 				results: make([]*blocksMetadata, 0, peersLen),
 			}
 			metadata[m.id.Hash()] = received
 		}
-		if received.submitted {
```

Comment:
I don't think we should move `received.submitted` up here. Honestly I don't know when this would happen, it seems like an error to me, but in any case if it's already submitted, we probably don't want to re-submit them